### ISSUE TITLE
Consideration of password locked names?

### ISSUE THREAD

**Issue by Jackster35** [Eureka22/agar-clone#16]
Have you considered password locking names? I understand people would be able to override it via ws://ip:port but maybe it could be incorporated in to the server too?

Just a thought :)

**Comment by Eureka22** [comment]
I will do multi server support later, will address this problem.

**Comment by Jackster35** [comment]
As so nickname locking would work? :dart:

**Comment by Eureka22** [comment]
Wait, what do you mean by nickname locking?

**Comment by Jackster35** [comment]
Hm, so certain names will require a password to be used. Like, "admin" would require password "123" in order to be used. I hope that makes more sense haha

**Comment by Eureka22** [comment]
That's more like registered users am I right? I will think about it but I am not sure if I will add it

**Comment by Jackster35** [comment]
Yeah, exactly what I was thinking of! But certain users instead of everybody.

**Comment by m-byte** [comment]
Since we are already using a modified Ogar-Server, it is possible. The easiest solution would have one password to unlock all otherwise locked names and a list of forbidden names. Could definitely be implemented. We would have to introduce some additional packets. Most likely one error packet (one or two byte(s) for an error code) and one password (MD5 or something) packet.

**Comment by vram4** [comment]
I always thought an account system would be cool. You could track stuff such as stats/achieves and friends lists. Maybe add a feature in which users that are registered would have different colored names than the unregistered ones?

**Comment by Jackster35** [comment]
That sounds like an awesome idea! Maybe link it to a mysql database or something and use PHP as the backend?

**Comment by vram4** [comment]
Yeah, i have experience with mysql but not with php unfortunately. Also if we were to make this, it should be in another repo as i feel that there might be some problems if i try to make Ogar compatible (mainly lag from extra code) with both clients.

**Comment by Jackster35** [comment]
Yeah, it would be a great idea as it'll make it easy to moderate abusive users, while keeping it clean. The statistic side would give some kind of competitiveness with it. Sounds like a lot to do so I don't expect you to do it, just ideas haha :)

**Comment by m-byte** [comment]
I would prefer to have everything based on node.js. That way, it could be one complete package. Makes it easier for users to install... Another reason against php: we need at least some kind of token in the websocket. It would definitely have to be integrated into Ogar.

Once Ogar-client is done, it would be possible to merge it with a fork (or maybe just a branch?) of Ogar. Then, a single npm install would suffice to give a user both server and client. We could call it edgar (EnhanceD oGAR) or cigar (Completely Integrated oGAR).

**Comment by m-byte** [comment]
Also, node.js would allow us to use passport for authentication. That way, almost any form of authentication is supported without us having to worry about it.

**Comment by m-byte** [comment]
I just realized, that having a user management would also allow for another feature:
All the administrative functions that are usually only available on the console could be redirected to the chat for users with the appropriate permissions.

**Comment by Eureka22** [comment]
@m-byte That would be awesome, we should seriously consider adding the user management.

**Comment by m-byte** [comment]
@Eureka22 Once Ogar-client is working ;) Shouldn't be too long from now, though. All the websocket stuff is already working and I am currently working on the drawing functions.

**Comment by Eureka22** [comment]
@m-byte Oh, it comes to me that when you write the drawing functions, are you going to use something other than canvas? Maybe some advanced render engines using webGL? That will make the graphics better and less demanding for hardware capacity. (As @vram4 his laptop cannot run official agar smoothly lol)

**Comment by m-byte** [comment]
Good idea. I already started the with canvas, but I'll search for a better renderer now and try to use that.

**Comment by Eureka22** [comment]
`pixi.js` might be a good choice. You can also find some here https://github.com/showcases/javascript-game-engines

**Comment by m-byte** [comment]
Ok, I'll switch from merge-js to browserify. That's probably the easiest way of managing the code, now that pixi will be used ;)

**Comment by vram4** [comment]
@m-byte We should just create a new repo for all this custom stuff.

**Comment by Eureka22** [comment]
@vram4 I can help manage, what about calling it CIGar :)

**Comment by m-byte** [comment]
Here you guys go: https://github.com/m-byte/Cigar
@Eureka22 @vram4 I added you as collaborators, so you'll have direct access to the main repo.

**Comment by Eureka22** [comment]
@m-byte Awesome, as the Ogar-client is not finished yet, I will first put my client into it. And we gradually replace that with new client :)

**Comment by Jackster35** [comment]
You guys are just awesome! :]

**Comment by Eureka22** [comment]
Oh @m-byte As for user management, I would like to use mongoDB, is that OK?

**Comment by m-byte** [comment]
@Eureka22 I would like to use passport for user-management. Standard setup would be in combination with passport-local, which uses mongoDB.

I like vram4's idea: We could use brighter colors for authenticated clients.

I'll write up some ideas for the protocol extensions needed to authenticate users.

Btw: we should probably just change the leaderboard to a link to github, whenever an old client connects.

**Comment by Eureka22** [comment]
@m-byte passport is easier to setup, better idea :)

As for the leaderboard, by doing that, do you mean you want the old client users to switch to new clients?

**Comment by m-byte** [comment]
Actually, thinking about it: we should support the official client as long as possible. Up to now, there is no difference between them that would cause vanilla clients to not work with it. So, we'll leave it as is. Once the support is broken, we can think about this.

**Comment by Eureka22** [comment]
@m-byte Yes. Maybe the official server will be adding things in and we need to change accordingly, but not a problem for now.

**Comment by m-byte** [comment]
My proposal for the new packets: https://github.com/m-byte/Cigar/issues/4
